lib/system/cellseqs_v1.nim

```diff
@@ -0,0 +1,43 @@
+#
+#
+#            Nim's Runtime Library
+#        (c) Copyright 2019 Andreas Rumpf
+#
+#    See the file "copying.txt", included in this
+#    distribution, for details about the copyright.
+#
+
+# ------------------- cell seq handling ---------------------------------------
+
+type
+  PCellArray = ptr UncheckedArray[PCell]
+  CellSeq {.final, pure.} = object
+    len, cap: int
+    d: PCellArray
+
+proc contains(s: CellSeq, c: PCell): bool {.inline.} =
+  for i in 0 .. s.len-1:
+    if s.d[i] == c: return true
+  return false
+
+proc add(s: var CellSeq, c: PCell) {.inline.} =
+  if s.len >= s.cap:
+    s.cap = s.cap * 3 div 2
+    var d = cast[PCellArray](alloc(s.cap * sizeof(PCell)))
+    copyMem(d, s.d, s.len * sizeof(PCell))
+    dealloc(s.d)
+    s.d = d
+    # XXX: realloc?
```

Comment:
this pattern is frequent, and often misses the realloc opportunity; strings and seq suffer from the same problem.

how about introducing a helper to make it easier to use everywhere needed:
```nim
proc reallocMaybe[T: ptr](a: T, n: num, fresh: var bool): T =
  ## for platforms/backends that don't support realloc, use copyMem + dealloc
  ## otherwise call b = realloc(a, n * sizeof(T))
  ## in either case, returns `fresh = true` if result != a
```

refs https://github.com/timotheecour/Nim/issues/239

Reply:
1. This is old code that moved around, not new code.
2. The capacity logic ensures we don't copy too often anyway, we don't need ``realloc`` for it.
3. In most cases ``realloc`` cannot enlarge the underlying block anyway and has to copy.